### PR TITLE
Op overload no fun variant

### PR DESCRIPTION
**Pull request by ajreynol**
Adds the ability of whether or not to accept function variants with overloading.
Changes the default behavior to not allow function variants.

### CODE REVIEW

File: src/expr/symbol_table.cpp [hunk 1]
```diff
@@ -205,29 +229,62 @@ bool OverloadedTypeTrie::markOverloaded(const string& name, Expr obj) {
...
-    // if there is already an active function with the same name and expects the
-    // same argument types
-    if (isOverloadedFunction(prev_obj)) {
-      if (prev_obj.getType() == obj.getType()) {
-        // types are identical, simply ignore it
-        return true;
-      } else {
-        // otherwise there is no way to distinguish these types, we return an
-        // error
+  if (d_allowFunctionVariants || argTypes.empty())
```
Comment (barrettcw) [inline src/expr/symbol_table.cpp]: So, if I understand correctly, when the flag is true, we allow functions to have the same name and argument types but to return different types.  We also always allow functions/constants with no arguments to have different types.  However, we never allow the redefinition of a function or constant with the same name, same arguments, and same return type as an existing definition.

When the flag is false, we don't allow function definitions with the same name and argument types, even if they return a different type.

Right?

File: src/expr/symbol_table.cpp [hunk 2]
```diff
@@ -205,29 +229,62 @@ bool OverloadedTypeTrie::markOverloaded(const string& name, Expr obj) {
...
+  {
+    // just check for redefinition
+    std::map<Type, Expr>::iterator it = tat->d_symbols.find(rangeType);
+    if (it != tat->d_symbols.end())
+    {
+      Expr prev_obj = it->second;
+      // if there is already an active function with the same name and expects
+      // the same argument types, we reject the re-declaration here
```
Comment (barrettcw) [inline src/expr/symbol_table.cpp]: *and* has the same return type

File: src/expr/symbol_table.cpp [hunk 3]
```diff
@@ -205,29 +229,62 @@ bool OverloadedTypeTrie::markOverloaded(const string& name, Expr obj) {
 
   // types can be identical but vary on the kind of the type, thus we must
   // distinguish based on this
```
Comment (barrettcw) [inline src/expr/symbol_table.cpp]: Can you expand the comment here to explain a bit more about what is happening in the code below - it's a bit tricky.

File: src/expr/symbol_table.cpp (hunk 1 above)
Comment (ajreynol) [reply src/expr/symbol_table.cpp]: Yes that's right. I've added further documentation in my latest commit to make this clear.

File: src/expr/symbol_table.cpp (hunk 2 above)
Comment (ajreynol) [reply src/expr/symbol_table.cpp]: Fixed

File: src/expr/symbol_table.cpp (hunk 3 above)
Comment (ajreynol) [reply src/expr/symbol_table.cpp]: I've improved it in the latest commit

**Comment by ajreynol** [comment]
I've addressed the review in my latest commit.

**Pull request merged**